### DISCUSSION TITLE
Deploying Security Onion with 3 Elasticsearch master-eligible nodes

### DISCUSSION THREAD

**Discussion by rwaight** [Security-Onion-Solutions/securityonion#5417] [General] [answered]
Is there a recommended way to deploy Security Onion with 3 master-eligible Elasticsearch nodes?  I am testing an SO environment with over 18 nodes and have been running into a `unhealthy manager` prompt when attempting to add more search nodes to my Security Onion grid (see https://github.com/Security-Onion-Solutions/securityonion/discussions/1751#discussioncomment-1288277 for reference).  This has me also thinking about the lack of High Availability (HA) with `TRUECLUSTER: true` (one of many checks [here](https://github.com/Security-Onion-Solutions/securityonion/blob/master/salt/elasticsearch/files/elasticsearch.yml#L4-L5)).  I'm hoping to leverage HA in Elasticsearch in the event the SO manager is ever restarted.

The reason I ask is based on this section of the [Elasticsearch 'voting-only master-eligible node' guide](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/modules-node.html#voting-only-node):

> High availability (HA) clusters require at least three master-eligible nodes, at least two of which are not voting-only nodes. Such a cluster will be able to elect a master node even if one of the nodes fails.
>
> Since voting-only nodes never act as the cluster’s elected master, they may require less heap and a less powerful CPU than the true master nodes. However all master-eligible nodes, including voting-only nodes, require reasonably fast persistent storage and a reliable and low-latency network connection to the rest of the cluster, since they are on the critical path for [publishing cluster state updates](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/cluster-state-publishing.html).

I was considering deploying 2 additional VMs as dedicated master-eligible ES nodes and connect them to the cluster with a custom firewall rule (to allow port 9300 between nodes) and custom certificates, but I do not want to go down an unsustainable path of customization to the current SO deployment configurations.

If there is not a recommended way to deploy SO with 3 master-eligible Elasticsearch nodes, would you consider an option to deploy an `Other` node configuration (from the install menu) for a [dedicated master-eligible Elasticsearch node](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/modules-node.html#dedicated-master-node)?  Ideally, this would include logic to limit the number of master-eligible nodes to 3 (not [voting-only nodes](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#voting-only-node)), although there is a [mitigation plan](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/modules-discovery-voting.html#_even_numbers_of_master_eligible_nodes) if there is an odd number.

Elasticsearch docs for reference:
* [Designing for resilience](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/high-availability-cluster-design.html#high-availability-cluster-design)
* [Bootstraping a cluster](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/modules-discovery-bootstrap-cluster.html#modules-discovery-bootstrap-cluster)
* [Setting the initial voting configuration](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/modules-discovery-voting.html#_setting_the_initial_voting_configuration)

For context, I saw [this question](https://github.com/Security-Onion-Solutions/securityonion/discussions/1751) about multiple managers as well as [this issue](https://github.com/Security-Onion-Solutions/securityonion/issues/80) regarding high availability within an SO deployment.

**Comment by TOoSmOotH** [comment, marked as answer]
If you look at the current logic in the elasticsearch.yml you can change the roles by modifying the minion file for that node and setting a custom role. So you could take an existing search node and assign it the master role in the minion file. 

From the elasticsearch.yml:

`{%- set NODE_ROLES = salt['pillar.get']('elasticsearch:node_roles', ['data', 'ingest']) %}`

So you would just need to add the additional role to it in the minion pillar for an existing search node:

```
elasticsearch:
  node_roles:
    - data
    - ingest
    - master
```

**Reply by rwaight** [reply]
That worked!  Thank you very much, @TOoSmOotH!
